### ISSUE TITLE
StorageStrategy - removeAtIndex add entity to parameters

### ISSUE THREAD

**Issue by damianroszyk** [angelnikolov/ts-cacheable#98]
Hello, 
thanks for the great lib!

I am writing Indexeddb IAsyncStorageStrategy and decided to store cache results in separate rows by changing cacheKey dynamically based on parameters using this method
```
  private static makeCacheKey(cacheKey: string, entity: ICachePair<any>): string {
    return `${cacheKey}#${MD5(JSON.stringify(entity.parameters))}`;
  }
```
This works great for me but I have encountered a problem here -> https://github.com/angelnikolov/ts-cacheable/blob/master/cacheable.decorator.ts#L67
`removeAtIndex` does not pass `entity` therefore I cannot use changed cacheKey to find it.
Any idea if we can add it there or if not why ?

Other solution I see (maybe even better) would be to pass method here -> 
https://github.com/angelnikolov/ts-cacheable/blob/master/cacheable.decorator.ts#L14
so we can change interface here to allow function returning string -> https://github.com/angelnikolov/ts-cacheable/blob/5561ed9f71b868e5d397d2e3ee52b82c952367dc/common/ICacheConfig.ts#L49
This way I can use cacheKey created dynamically in all methods from `IAsyncStorageStrategy`

I would like to know what you think this ( and create PR if you approve change :) )

**Comment by angelnikolov** [comment]
@damianroszyk Sorry for the late response, but can you provide a snippet of where you use `makeCacheKey` and where you expect `entity` to be provided? Also, is entity the found cache pair?

**Comment by damianroszyk** [comment]
Hi, sorry for my late response :) 
Here is the strategy that I wrote ( relevant parts ) that uses `idb` lib to store cache in indexeddb 
```
export class IdbStrategy<Db extends DBSchema> extends IAsyncStorageStrategy {

  private makeCacheKey(cacheKey: string, entity: ICachePair<any>) {
    return `${cacheKey}#${MD5(JSON.stringify(entity.parameters))}` as StoreKey<
      Db,
      any
    >;
  }

  async add(
    entity: ICachePair<any> & StoreValue<Db, any>,
    cacheKey: string | StoreKey<Db, any>
  ): Promise<void> {
    if (await isPrivateMode()) {
      return null;
    }
    await (await this.database).put(
      this.storeName,
      entity,
      this.makeCacheKey(cacheKey as string, entity)
    );
  }

  async removeAtIndex(index: number, cacheKey: string): Promise<void> {
    // no entity so we cannot makeCacheKey
    // return await (await this.database).delete(this.storeName', IdbStrategy.makeCacheKey(cacheKey, null));
  }
}
``` 

Here is the ICachePair interface which I use to create cacheKey
```
export interface ICachePair<T> {
  parameters: any;
  response: T;
  created: Date;
```

**Comment by angelnikolov** [comment]
@damianroszyk Hm, I got your idea. Maybe a `remove` method would make a lot more sense, but I remember there was a reason to remove by index. We have a couple of options:
1. Deprecate `removeByIndex` and introduce `remove` (and update our internal storage strategies to use it instead)
2. Add the extra entity parameter to the callback (what you suggested)

In your case, what would be a better and cleaner option?

**Comment by damianroszyk** [comment]
In my case I just need extra parameter `entity` and that'd be sufficient, but then I'd think about changing `updateAtIndex` the same way...
So maybe for now it's better to equalize those two methods

```
  abstract updateAtIndex(index: number, entity: ICachePair<any>, cacheKey: string): void | Promise<void>;
  abstract removeAtIndex(index: number, cacheKey: string): void | Promise<void>;

-->

  abstract updateAtIndex(index: number, entity: ICachePair<any>, cacheKey: string): void | Promise<void>;
  abstract removeAtIndex(index: number, entity: ICachePair<any>, cacheKey: string): void | Promise<void>;
```

wdyt ?
However it'd be a breaking change because of parameters order so I am not sure what's best here.

**Comment by angelnikolov** [comment]
To be honest, I am more inclined to add the `update()` and `remove()` methods and use those with priority instead of the `index`-based ones. Would that work for you?

**Comment by damianroszyk** [comment]
Sounds good!

**Comment by angelnikolov** [comment]
Fixed in #99
